### PR TITLE
Modernise buttons across Jenkins

### PR DESCRIPTION
**Pull request by janfaracik**
#### Screenshots 

image

image

---

#### What's new

This PR modernises the buttons across Jenkins, adding a consistent styling for all buttons inline with the new forms design. 

This PR also takes the first steps in removing dependency on YUI, enabling a leaner and more modern frontend that's easier to develop and maintain. 

---

#### Questions and answers

* "Why `.jenkins-button`? Why not just restyle `button`?"
  Restyling `button` would impact every button across Jenkins and could have negative consequences, creating a new class negates that. 
  Styling base elements can have unwanted effects for developers/plugin creators and could clash with their own styling.
  Adding a new class also enables us to style links to appear like buttons should we so wish.
  Having ownership of our styling makes it easy to update in the future and has no dependency on others (such as YUI/Bootstrap).

* "This PR breaks plugin X"
  Post a comment and I'll try my best to get the plugin working properly with the new styling/YUI cut down :)

---

#### Test this pull request

```command
docker run --rm -ti -p 8080:8080 -e ID=5897 -e MERGE_WITH=master jenkins/core-pr-tester
```

---

### Proposed changelog entries

* Modernise buttons across Jenkins

### Proposed upgrade guidelines

* For plugin developers, use `<button/>` instead of `<input/>` for buttons, add the new `jenkins-button` class for styling along with any modifiers that you may need (`--primary`, `--transparent`, `--destructive`).

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
Should this be in draft status? See "Convert to draft" in the sidebar.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (timja) [inline war/src/main/webapp/scripts/hudson-behavior.js:687]: 👍 used in a few plugins unfortunately: https://github.com/search?q=org%3Ajenkinsci+makeButton&type=code

**Review by timja — COMMENTED** [review]
Wow!

Very nice work, I've gone through and tested most of the buttons that were changed.

~Repeatable delete and api token issues are the blockers~

~Git plugin broken is a blocker~ https://github.com/jenkinsci/git-plugin/pull/1171

~The tests are failing because of markup changes~

 We will need to give this a run through ATH and bom PCT as well so that it doesn't break a lot of tests when merging

**Comment by timja** [comment]
> For plugin developers, use <button/> instead of <input/> for buttons, add the new jenkins-button class for styling along with any modifiers that you may need (--primary, --transparent, --destructive).

Stick that and maybe some samples in https://www.jenkins.io/doc/developer/views/ ?

Something like storybook would be better but that's likely a lot of work

**Comment by timja** [comment]
Git plugin configuration is broken:

image

The add and delete buttons don't work along with looking broken

File: core/src/main/resources/lib/form/repeatable/repeatable.js
Comment (timja) [inline core/src/main/resources/lib/form/repeatable/repeatable.js:166]: this breaks compatibility

**Comment by timja** [comment]
Code seems to be here:
https://github.com/jenkinsci/git-plugin/blob/5edc26da3bd4962bdb8599ef1cbc40d335cd1129/src/main/resources/hudson/plugins/git/UserRemoteConfig/config.groovy#L31-L32

I assume it needs a class added, (`jenkins-button`?)

**Comment by timja** [comment]
Sign in button:
image

**Comment by janfaracik** [comment]
> > For plugin developers, use  instead of  for buttons, add the new jenkins-button class for styling along with any modifiers that you may need (--primary, --transparent, --destructive).
> 
> Stick that and maybe some samples in https://www.jenkins.io/doc/developer/views/ ?
> 
> Something like storybook would be better but that's likely a lot of work

Will do! :)

**Comment by janfaracik** [comment]
> Sign in button:
> image

Good spot! Having a quick browse through the login/signup stuff - few questions:

1. Is there a reason we can't use .less for those pages? 
2. Are the font-awesome/google-fonts/icomoon folders used anywhere?

Thanks :)

**Comment by timja** [comment]
> Is there a reason we can't use .less for those pages?

none except for those files are intentionally separate from the rest of the styles / layout so that they are absolute minimal attack surface, so fine to share something like variables and button styles, but not all css for the application probably.

> Are the font-awesome/google-fonts/icomoon folders used anywhere?

Probably not but not sure

File: war/src/main/less/modules/buttons.less
Comment (fqueiruga) [inline war/src/main/less/modules/buttons.less:112]: Should come from a CSS var

**Review by fqueiruga — COMMENTED** [review]
I have several thoughts about this PR:

- The buttons are too deeply tied to YUI and I'm afraid that this could become a mini tables-to-divs.
- I understand the need to simplify design and configuration, but just leaving the option to customize the primary button background is too mucho IMO. At the very least, all backgrounds and colours should be configurable.
    - Could you please also provide screenshots for the hover & focus states?
- I wonder if the contrast for the danger button is accessible enough.

**Comment by timja** [comment]
Administrative monitors have no spacing:

Before:
image


After:
image

**Comment by timja** [comment]
Expandable text area doesn't take up full line:

Freestyle build => Add built step => Invoke maven top level targets

Before:
image


After:
image

**Comment by timja** [comment]
> The buttons are too deeply tied to YUI and I'm afraid that this could become a mini tables-to-divs.

Would suggest trying it out if you haven't already, my test setup (which has a lot of plugins in it) only has 2 serious rendering issues:
* the git plugin which I'll look at soon.
* sign in page login button (probably will leave that to Jan)

**Comment by timja** [comment]
Credentials yui button looks weird:

image

**Comment by timja** [comment]
ATH smoke test compat: https://github.com/jenkinsci/acceptance-test-harness/pull/709

**Comment by fqueiruga** [comment]
I have found a few issues:

Form not expanding when it should, probably due to a missing call of `layoutUpdateCallback`, maybe has to do with the removal of `expandButton`:

https://user-images.githubusercontent.com/5738588/140901611-303a3ba5-959a-45b0-9e1a-7f15b015a191.mov

JDK configuration is broken, maybe has to do with the changes at repeatable:

https://user-images.githubusercontent.com/5738588/140901677-fb0490f6-4184-4d68-bd0d-58f281244256.mov

IMO this one should have a big compatibility layer baked in. We did it for tables to divs and I think these changes are serious enough to warrant it.

**Review by uhafner — COMMENTED** [review]
I think these changes are still a little bit fragile for a detailed review. I'm also not sure if changing the buttons is so important since the buttons already have been restyled nicely by Felix and Joseph. I think it would make more sense to finish the form restyling first before we head to another thing that potentially breaks existing forms. 

Some comments for the current state (I do not comment on things that do not work currently).
- icon-button: the horizontal distribution of the icon and the text looks somehow misaligned. The icons should move some more pixels to the left. Here an example of how it could look: https://www.codehim.com/wp-content/uploads/2020/08/bootstrap-5-buttons-with-icon-and-text.png 
Bildschirmfoto 2021-11-09 um 11 50 18. 
- While I understand that is is helpful to use a consistent styling I must admit that I liked the old repeatable delete buttons with the simple "X" much better than a button with the text *delete*. But maybe it is caused by the wrong alignment in the current version of this PR:
Bildschirmfoto 2021-11-09 um 11 55 04

 - As Felix noted, the repeatable add is broken right now. Adding new dialog elements dynamically based on the selection does not work. Maybe that is the reason that all ATH tests fail. 
 - The modified state of the section below the advanced button should be integrated somehow into the button itself. 
Bildschirmfoto 2021-11-09 um 11 59 29

**Comment by timja** [comment]
> Maybe that is the reason that all ATH tests fail

It's not, ATH tests fail because of selectors / markup changes and form-element-path plugin not working for buttons like this (it's quite a hacky plugin), the smoke tests are fixed in this PR: https://github.com/jenkinsci/acceptance-test-harness/pull/709

It's got all tests enabled atm, I'm hoping that form-element-path fixes will remove a lot of the issues but we'll see, right now there's too many failing 😢

File: core/src/main/resources/lib/form/repeatable/repeatable.js
Comment (timja) [inline core/src/main/resources/lib/form/repeatable/repeatable.js:172]: we can just add comma's between selectors apparently rather than duplicating: https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/scripts/hudson-behavior.js#L1213

**Review by basil — COMMENTED** [review]
Have we verified that this will not cause regressions in the weekly BOM line?

**Comment by timja** [comment]
> Have we verified that this will not cause regressions in the weekly BOM line?

I’m working on getting ATH green first, so that an incremental will deploy.

but yes will do as I said here: 
https://github.com/jenkinsci/jenkins/pull/5897#pullrequestreview-799505489

File: core/src/main/resources/lib/form/repeatable/repeatable.js
Comment (timja) [inline core/src/main/resources/lib/form/repeatable/repeatable.js:170]: not sure where exactly the issue is but the 'path' generation isn't working right for hetero elements.

Steps:

1. http://localhost:8080/script
2. `jenkins.formelementpath.FormElementPathPageDecorator.ENABLED = true`
3. Create new freestyle job
4. Add post-build action such as archive artifacts

Check path of post-build artifacts button by inspecting the button:

expected:
`/hetero-list-add[publisher]`

actual:
`/publisher[ArtifactArchiver]/hetero-list-add[publisher]`

This can be reproduced with this test: https://github.com/jenkinsci/acceptance-test-harness/blob/01bd5840273170bbe3c6c7c82aa9479acb5181c1/src/test/java/core/FreestyleJobTest.java#L234

Comment (janfaracik) [reply core/src/main/resources/lib/form/repeatable/repeatable.js:170]: Should be fixed now - `hetero-list-add` still uses YUI so needed to update the condition to reflect that :)

**Comment by janfaracik** [comment]
> Expandable text area doesn't take up full line:
> 
> Freestyle build => Add built step => Invoke maven top level targets
> 
> Before: image
> 
> After: image

Updated :)

image

**Comment by janfaracik** [comment]
Thanks for the review!

> * icon-button: the horizontal distribution of the icon and the text looks somehow misaligned. The icons should move some more pixels to the left. Here an example of how it could look: https://www.codehim.com/wp-content/uploads/2020/08/bootstrap-5-buttons-with-icon-and-text.png

I agree :) I've adjusted the spacing -

image

>   Bildschirmfoto 2021-11-09 um 11 50 18.
> * While I understand that is is helpful to use a consistent styling I must admit that I liked the old repeatable delete buttons with the simple "X" much better than a button with the text _delete_. But maybe it is caused by the wrong alignment in the current version of this PR:

Also agree, I've got that updated in #5923, no picture of it on hand but its a little red circle with an X in it :)

> Bildschirmfoto 2021-11-09 um 11 55 04
> 
> * As Felix noted, the repeatable add is broken right now. Adding new dialog elements dynamically based on the selection does not work. Maybe that is the reason that all ATH tests fail.

Think that was an issue due to the original forms update PR that got merged, all fixed now thanks to @timja :)

> * The modified state of the section below the advanced button should be integrated somehow into the button itself.
> 
> Bildschirmfoto 2021-11-09 um 11 59 29

That's a good idea! I'll look into it and see what I can do :) 

Thanks again

File: core/src/main/resources/lib/form/repeatable/repeatable.js
Comment (timja) [reply core/src/main/resources/lib/form/repeatable/repeatable.js:170]: yeah that fixed it

**Comment by timja** [comment]
> Also agree, I've got that updated in #5923, no picture of it on hand but its a little red circle with an X in it :)

picture of what the delete button looks like in form re-work:

image

File: core/src/main/resources/lib/form/advanced.jelly
Comment (timja) [inline core/src/main/resources/lib/form/advanced.jelly:52]: Is there a reason this is at the end of the input whereas others have it at the start?

Comment (janfaracik) [reply core/src/main/resources/lib/form/advanced.jelly:51]: Initial thinking behind putting it on the right was that it's an addition to the button in the case that any fields have been edited, and not visible in every case. Not too sure how I feel about its positioning, I've moved it back to the left side so its consistent with other icons.

Comment (timja) [reply core/src/main/resources/lib/form/advanced.jelly:51]: ahk it's a bit weird anyway

Comment (daniel-beck) [reply core/src/main/resources/lib/form/advanced.jelly:51]: Isn't an asterisk a bit too similar to various decorations indicating the creation of something new? E.g. the glare/sun highlights for new document etc. icons in Tango: https://commons.wikimedia.org/wiki/Tango_icons (until I looked it up I was confident it was an actual asterisk!)

Unsure what else to use though. Perhaps something more attention grabbing, like `!` or a similar warning?

Comment (timja) [reply core/src/main/resources/lib/form/advanced.jelly:51]: do people actually know it means something (the symbol inside advanced)? I wasn't aware of what it meant

Comment (janfaracik) [reply core/src/main/resources/lib/form/advanced.jelly:51]: Ah an asterisk is commonly used for when something isn't saved, eg in Notepad, and that was the closest thing I could think of (although our situation is different)

image

I'm thinking `!` might be a little too attention grabbing? Wouldn't want the user to think they've done something wrong - it's a hard one

Comment (pcarles) [reply core/src/main/resources/lib/form/advanced.jelly:51]: what about a wrench or a gear on the left?

**Comment by timja** [comment]
conflicted

**Review by NotMyFault — CHANGES_REQUESTED** [review]
A few things I noticed while giving this a test:

- The button text on `/prepareShutdown` doesn't fit into the button

The same applies to the button to cancel the shutdown


- The button at your profile to terminate active sessions doesn't fit into any button scheme, as proposed on the PR description, yet it 


- The build history table has some styling issues with different themes:


- The same applies to the icon size buttons

The font color doesn't match to the font color of the job items in the table view as well.

- Was https://github.com/jenkinsci/jenkins/pull/5897#issuecomment-962628558 ever addressed? It applies (to me) still.

- The text on the default, red and icon button have some sort of background text that doesn't work well with different themes
<details>
<summary>Button theme issues</summary>



</details>

<details>
<summary>Other theme issues</summary>



</details>

- Smaller buttons in the job configuration have a hard time with the new buttons as well.


These ones are from the credentials plugins for example.

**Comment by timja** [comment]
conflicts

**Comment by timja** [comment]
conflicts 😄 

quite a few tests and ATH failing too.

**Comment by timja** [comment]
conflicts ^.^

**Comment by timja** [comment]
Nice tests are passing, on to ATH :)

File: core/src/main/resources/hudson/PluginWrapper/uninstall.groovy
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginWrapper/uninstall.groovy:19]: Is the expectation here really that plugin developers copy & paste code like this into their plugin so it integrates nicely with the UI? Shouldn't this kind of thing be offered as a reusable element, perhaps with custom attributes that are easy to understand and internally translate it into whatever CSS classes you want?

Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginWrapper/uninstall.groovy:19]: Is the expectation here really that plugin developers copy & paste code like this into their plugin so it integrates nicely with the UI? Shouldn't this kind of thing be offered as a reusable element, perhaps with custom attributes that are easy to understand and internally translate it into whatever CSS classes you want?

File: core/src/main/resources/hudson/model/ComputerSet/index.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/ComputerSet/index.jelly:37]: Similarly, this seems actively hostile to plugin developers to not have a Jelly control.

Also, I still don't understand the naming convention. Do we document anywhere when to use `-`, `--`, `__`, or `-!-` as separator?

File: core/src/main/resources/hudson/model/Job/configure.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Job/configure.jelly:63]: What happens in views that currently set this attribute? Is it just an obsolete no-op?

File: core/src/main/resources/jenkins/security/seed/UserSeedProperty/config.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/security/seed/UserSeedProperty/config.jelly:12]: What happens in plugins that currently use this?

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:1228]: Is this a compatible change with these buttons in plugins?

File: war/src/main/js/widgets/config/tabbar.less
Comment (daniel-beck) [inline war/src/main/js/widgets/config/tabbar.less:41]: Unrelated?

File: core/src/main/resources/lib/layout/expandButton.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/expandButton.jelly:1]: 1 plugin uses this per https://github.com/search?q=org%3Ajenkinsci+expandbutton&type=Code and it's suspended: https://plugins.jenkins.io/javanet/

GitHub code search is known incomplete though, did you check all plugins' resources? This also is a breaking change that needs to be documented.

File: core/src/main/resources/lib/hudson/iconSize.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/hudson/iconSize.jelly:40]: Should be a parameterized localizable expression to account for different grammars.

**Comment by timja** [comment]
conflicts

File: core/src/main/resources/lib/layout/expandButton.jelly
Comment (timja) [reply core/src/main/resources/lib/layout/expandButton.jelly:1]: Nothing extra in the new code search which is a lot better: https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Ajenkinsci+expandButton

**Comment by timja** [comment]
> ### Proposed upgrade guidelines
> * For plugin developers, use `<button/>` instead of `<input/>` for buttons, add the new `jenkins-button` class for styling along with any modifiers that you may need (`--primary`, `--transparent`, `--destructive`).

Possible to document this either on jenkins.io or in ui samples?

**Comment by timja** [comment]
Developer docs pushed to ui samples plugin:
image

**Comment by alecharp** [comment]
@janfaracik it seems that we have new conflicts on this PR and looking that the ATH failure, they seem to be related to the changeset. I guess the tests need to be updated to the new button you are offering here.

File: core/src/main/resources/hudson/PluginWrapper/uninstall.groovy
Comment (janfaracik) [reply core/src/main/resources/hudson/PluginWrapper/uninstall.groovy:19]: IMO a standard button is too simple to be packaged as a component, its just a combination of its content/class/click event. If the button had other functionality, eg animations/complex states/UI, then I'd definitely be in support of making it its own component, such as the help button component.

Happy to be convinced otherwise though/I'm good either way 👍

File: core/src/main/resources/hudson/model/Job/configure.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/model/Job/configure.jelly:63]: The attribute is ignored 👍

**Comment by timja** [comment]
You can just run `mvn deploy` from the war directory btw if you want a quick snapshot to run in the ATH pipeline

**Comment by janfaracik** [comment]
Heya folks. 

Wanting to get some opinions on this, should I continue with this PR and fix up the tests or should I close this PR (keeping the branch) and instead post multiple (eg 5+) lightweight PRs based off of this (such as "Update bottom bar buttons", "Update dropdown buttons", "Update RSS buttons")? 

Just wanting to make life easier for everyone.

**Comment by NotMyFault** [comment]
Separating the work into several smaller PRs is definitely the +1 choice for me for the reasons basil outlined on your polishing PR.

**Comment by timja** [comment]
> Wanting to get some opinions on this, should I continue with this PR and fix up the tests or should I close this PR (keeping the branch) and instead post multiple (eg 5+) lightweight PRs based off of this (such as "Update bottom bar buttons", "Update dropdown buttons", "Update RSS buttons")?

Smaller PRs would be preferred.

At this point of the PR not sure though, it's the ATH failures that are holding it back.

**Comment by janfaracik** [comment]
> > Wanting to get some opinions on this, should I continue with this PR and fix up the tests or should I close this PR (keeping the branch) and instead post multiple (eg 5+) lightweight PRs based off of this (such as "Update bottom bar buttons", "Update dropdown buttons", "Update RSS buttons")?
> 
> Smaller PRs would be preferred.
> 
> At this point of the PR not sure though, it's the ATH failures that are holding it back.

That's true, just worried that once its merged in it could have unforeseen consequences like the breadcrumb one.

**Comment by basil** [comment]
Agreed that ATH failures should be the priority, but I still think it is worth splitting up large PRs like this as a risk reduction strategy. #6395 was self-contained enough that I could at least read (but not understand) the change in a few minutes, which enabled me to find the proximate cause of [JENKINS-68270](https://issues.jenkins.io/browse/JENKINS-68270) with about an hour of debugging, even though I had never seen the code before and had basically no understanding of it. I doubt I would have been able to do this had the change been much larger.

**Comment by janfaracik** [comment]
Thanks to all those that took a look and helped out, it's super appreciated. I'm going to close this PR and instead create a series of smaller ones. They'll be based off of this PR just in far smaller chunks (e.g. less than ten files each ideally), all the  contributions/fixes that helped to shape this piece of work will carry across to the smaller PRs so nothing is lost.

Thanks again, Jan.

**Pull request closed without merging**
